src/hotspot/share/classfile/javaClasses.cpp

```diff
@@ -2735,49 +2735,57 @@ void java_lang_Throwable::get_stack_trace_elements(int depth, Handle backtrace,
   }
 }
 
-Handle java_lang_Throwable::get_cause_with_stack_trace(Handle throwable, TRAPS) {
-  // Call to JVM to fill in the stack trace and clear declaringClassObject to
-  // not keep classes alive in the stack trace.
-  // call this:  public StackTraceElement[] getStackTrace()
+Handle java_lang_Throwable::create_initialization_error(JavaThread* current, Handle throwable) {
+  // Creates an ExceptionInInitializerError to be recorded as the initialization error when class initialization
+  // failed due to the passed in 'throwable'. We cannot save 'throwable' directly due to issues with keeping alive
+  // all objects referenced via its stacktrace. So instead we save a new EIIE instance, with the same message and
+  // symbolic stacktrace of 'throwable'.
   assert(throwable.not_null(), "shouldn't be");
 
-  JavaValue result(T_ARRAY);
-  JavaCalls::call_virtual(&result, throwable,
-                          vmClasses::Throwable_klass(),
-                          vmSymbols::getStackTrace_name(),
-                          vmSymbols::getStackTrace_signature(),
-                          CHECK_NH);
-  Handle stack_trace(THREAD, result.get_oop());
-  assert(stack_trace->is_objArray(), "Should be an array");
-
-  // Throw ExceptionInInitializerError as the cause with this exception in
-  // the message and stack trace.
-
-  // Now create the message with the original exception and thread name.
+  // Now create the message from the original exception and thread name.
   Symbol* message = java_lang_Throwable::detail_message(throwable());
-  ResourceMark rm(THREAD);
+  ResourceMark rm(current);
   stringStream st;
   st.print("Exception %s%s ", throwable()->klass()->name()->as_klass_external_name(),
              message == nullptr ? "" : ":");
   if (message == nullptr) {
-    st.print("[in thread \"%s\"]", THREAD->name());
+    st.print("[in thread \"%s\"]", current->name());
   } else {
-    st.print("%s [in thread \"%s\"]", message->as_C_string(), THREAD->name());
+    st.print("%s [in thread \"%s\"]", message->as_C_string(), current->name());
   }
 
   Symbol* exception_name = vmSymbols::java_lang_ExceptionInInitializerError();
-  Handle h_cause = Exceptions::new_exception(THREAD, exception_name, st.as_string());
-
-  // If new_exception returns a different exception while creating the exception, return null.
-  if (h_cause->klass()->name() != exception_name) {
+  Handle h_eiie = Exceptions::new_exception(current, exception_name, st.as_string());
```

Comment:
I don't like the h_eiie name.  I keep trying to pronounce it in English.  How about "initialization_error" ?